# first, find matches to filenames of this form:
# Load sampleNames from the config.yaml file
configfile: "workflow/config/config.yaml"
sampleNames = config["sampleNames"]
projectName = config["projectName"]
megahit_path = Path("results/megahit/assembly")
megahit_assembly_fa = megahit_path / f"{projectName}_final.contigs.fa"
megahit_orfs_ffn_fname = megahit_path / f"{projectName}_final_orfs.ffn"


include: "rules/Snakefile1GetPrepReadsRibodetector.smk"
include: "rules/Snakefile2Assemble.smk"
include: "rules/Snakefile3.0SalmonIndex.smk"
include: "rules/Snakefile3MapReads.smk"

rule all:
    input:
        # expand("results/intermediate_fastqs/{sra_id}_filtered.fq.gz", sra_id=sampleNames),
        # expand("results/intermediate_fastqs/{sra_id}_aligned.fq.gz", sra_id=sampleNames),
        expand("results/salmon/{sra_id}/quant.sf", sra_id=sampleNames),
        "results/megahit/assembly/final.contigs.fa",
        megahit_assembly_fa,
        megahit_orfs_ffn_fname,
        "results/salmon/transcriptome_index/seq.bin",
        "logs/salmon/transcriptome_index.log",
